### PR TITLE
Stream overhaul

### PR DESCRIPTION
**Pull request by ErlendHaa**

### CODE REVIEW

File: lib/include/dlisio/io.hpp [hunk 1]
```diff
@@ -21,7 +21,7 @@ namespace dl {
  */
 class stream {
 public:
-    explicit stream( lfp_protocol* p ) noexcept (false);
+    explicit stream( lfp_protocol* p ) : f(p) {};
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply lib/include/dlisio/io.hpp]: 1. http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c49-prefer-initialization-to-assignment-in-constructors
2. / 3. It's noexcept (true)

File: lib/include/dlisio/stream.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    explicit stream( lfp_protocol* p ) : f(p) {};
+
+    lfp_protocol* protocol() const noexcept (true);
+
+    /* Seek within the logical domain of the current file. I.e. seeks are
+     * performed with the outer-most lfp_protocol. */
+    void lseek( std::int64_t offset ) noexcept (false);
+
+    /* The tell reported by the current (outer-most) lfp_protocol */
+    std::int64_t ltell() const noexcept(true);
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply lib/include/dlisio/stream.hpp:24]: Only thing missing is cstdint, will add

File: lib/src/io.cpp [hunk 3]
```diff
@@ -218,7 +218,7 @@ noexcept (false) {
 
 }
 
-void stream::close() {
+void stream::close() noexcept (true) {
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply lib/src/io.cpp]: Unsure what those are - and if we care in dlisio. But untill we do anything about it, it's noexcept (true)

File: lib/src/io.cpp [hunk 4]
```diff
@@ -242,13 +242,13 @@ void stream::seek( std::int64_t offset ) noexcept (false) {
     }
 }
 
-std::int64_t stream::tell() const noexcept (true) {
+std::int64_t stream::ltell() const noexcept (true) {
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply lib/src/io.cpp]: Agree

File: lib/include/dlisio/io.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (achaikou) [reply lib/include/dlisio/io.hpp]: 1. I don't get something. `f(p) {}` is good. But why in .hpp, not in .cpp?
2. So you are going to add it? 

> Omitting the noexcept specifier is equivalent to noexcept(false)

File: lib/include/dlisio/stream.hpp [hunk 5]
```diff
@@ -17,8 +17,39 @@ class stream {
...
+     *                        ^                 ^
+     *                        |                 |
+     *                  lfp_protocol         current
+     *                     opened            position
+     *                      here             in file
+     *
+     * In the above example where the current protocol was open at ptell=100
+     * and that the 200 bytes corresponding to some header are abstracted away
+     * by the lfp_protocol, and the current position in the file is as marked.
+     * Then ltell == 150 and ptell = The ltell
```
Comment (achaikou) [inline lib/include/dlisio/stream.hpp:56]: 1. ptell = ??
2. Actually the paragraph structure is weird:
    In the above example where X and that Y, and Z. Then RESULT.
  More sense would be:
   In the above example X, Y, and Z. This makes RESULT.

File: lib/include/dlisio/stream.hpp [hunk 6]
```diff
@@ -7,10 +7,16 @@
 
 namespace dl {
 
-/* Stream - wrapper for lfp_protocol
+/* Stream - C++ wrapper for lfp/lfp_protocol
```
Comment (achaikou) [inline lib/include/dlisio/stream.hpp:10]: Meaning of / ?

File: lib/include/dlisio/stream.hpp [hunk 7]
```diff
@@ -7,10 +7,16 @@
...
  *
- * The main purpose of stream is to handle lfp return codes in a manner that
- * suits dlisio and make a more ergonomic interface for caller functions.
+ * The main purpose of stream is to wrap lfp's C API into something more
+ * C++-ish. lfp return codes are translated into exceptions. All functions
+ * except ptell work on the logical file as presented by the current
+ * lfp_protocol. E.g. take a stream-instance with the protocol stack lfp_cfile,
+ * lfp_tapeimage and lfp_rp66 opened on a dlis-file. This stream will only
+ * "see" bytes corresponding to the Logical Record Segments. I.e. the file will
+ * look like a series of LRS' from start to finish. All bytes corresponding to
```
Comment (achaikou) [inline lib/include/dlisio/stream.hpp:18]: a series of LRS

' in the end is for possessives. If you want plural, but it is kinda strange - LRS'es? LRS's??
series of LRS should be enough I think

**Pull request merged**